### DISCUSSION TITLE
Load form values from local storage

### DISCUSSION THREAD

**Discussion by etomanon** [react-hook-form/react-hook-form#3206] [Q&A] [answered]
Hello,

I have problem with loading form values from local storage because it causes endless loop.
I guess that watch() is updated more frequently than on change?
Any idea how to save data to local storage (on form change) and then on mount (first render) load data from local storage to the form?
Here is my approach:
https://codesandbox.io/s/endless-loop-local-storage-rxpup

**Comment by bluebill1049** [comment, marked as answer]
i would read the local storage data at useEffect and use reset to update the entire form.

**Reply by etomanon** [reply]
@bluebill1049 
I tried it that way too but I still get endless re-render loop.
I believe the problem is with saving the form data to local storage.
I want to save each change to the local storage (not only on submit)
Any idea?
```
const formData = watch()
// save new data on form change to local storage
// probably endless loop???
useEffect(() => {
  if (formData) {
    setStorage(formData);
  }
}, [formData, setStorage]);

// load data from local storage
 useEffect(() => {
    if (!initialized && storage) {
      setInitialized(true);
      reset(storage);
    }
  }, [initialized, storage, reset]);
```
https://codesandbox.io/s/endless-loop-local-storage-forked-3udoh?file=/src/App.js:535-717

**Reply by bluebill1049** [reply]
I tried with just reset and it's working as expected: https://codesandbox.io/s/endless-loop-local-storage-forked-hiznj?file=/src/App.js

**Reply by etomanon** [reply]
@bluebill1049 
It is not working.
When you change the input value and then refresh the page (F5) you still see the old initial value (name initial).
I would like to change the input value to "name changed" and then (without submitting the form) when you refresh the page you would see the "name changed" value as default (initial) value.

**Reply by bluebill1049** [reply]
that's because in my example  data is not stored in the local stage, what i try to proof that `reset` works. 

- reset should only invoke once when the app mounted
- you need to store data into local storage by action or watch the entire form and update the local storage

**Reply by etomanon** [reply]
@bluebill1049 
That is exactly what I was doing in my 1st version but the `const formData = watch()`  causes infinite loop. 
This does not help either: `const formData = useMemo(() => watch(), [watch])`
So I guess there is no way to run useEffect only when a form value changes?
```
const formData = watch();
  // useEffect runs infinite loop - breaks app :(
  useEffect(() => {
    // console.log(formData);
    // setStorage(formData);
  }, [formData]);
```
https://codesandbox.io/s/endless-loop-local-storage-forked-zu7oc?file=/src/App.js

**Reply by bluebill1049** [reply]
i would attach onChange on your form. formState is not memories from watch, it was supposed to be using inside the render.

**Reply by ialsowalkdogs** [reply]
@etomanon 
Hello! Did you manage to solve your issue? I am working on a similar case and would be grateful if you shared a result that works for you.

**Reply by etomanon** [reply]
@berrybell 
Yes, I solved it: https://codesandbox.io/s/endless-loop-local-storage-forked-4dgu8?file=/src/App.js:312-319

**Reply by pkino** [reply]
@bluebill1049 
> i would read the local storage data at useEffect and use reset to update the entire form.

I have a question about this.
Which do you think is better reading local storage data at useEffect or reading local storage data at default values of useForm and firing trigger at useEffect.
In my case, I have a disabled submit button until validation is ok, so I have to validate local storage data.

**Reply by bluebill1049** [reply]
if you can, `defaultValues` at `useForm` seems to be a better option.

**Reply by chrift** [reply]
FWIW, if anyone else comes across this, I have started to adopt a pattern of having my page actually being a dataloader, which renders the main page content after it knows whether there are default values (loaded from localStorage/localforage)

It seemed to be more reliable than using a reset in a useEffect or reading from the storage actually in the defaultValues assignment.

Like this:

```

export default function PageDataLoader() {
  const [initialData, setInitialData] = useState<Form | null>();

  useEffect(() => {
    const loadFormData = async () => {
      const savedForm = await localforage.getItem<Form>(
        'storage-key'
      );

      setInitialData(savedForm);
    };

    loadFormData();
  }, []);

  if (initialData !== undefined) {
    return <Page initialData={initialData || undefined} />;
  }

  return 'Loading...';
}


function Page({
  initialData,
}: {
  initialData?: Form;
}) {
  const {
    control,
    handleSubmit,
    register: registerInput,
    watch,
    setValue,
  } = useForm<Form>({
    values: initialData,
    shouldUnregister: true,
  });
}
```
